File: src/main/java/org/apache/prepbuddy/datacleansers/imputation/MeanSubstitution.java
package org.apache.prepbuddy.datacleansers.imputation;

import org.apache.prepbuddy.rdds.TransformableRDD;
import org.apache.prepbuddy.utils.RowRecord;
import org.apache.spark.api.java.JavaDoubleRDD;

/**
 * An imputation strategy that imputes the missing column value with the
 * mean of the specified column.
 */
public class MeanSubstitution implements ImputationStrategy {

    private Double mean;

    @Override
    public void prepareSubstitute(TransformableRDD rdd, int missingDataColumn) {
        JavaDoubleRDD javaDoubleRDD = rdd.toDoubleRDD(missingDataColumn);
        mean = javaDoubleRDD.mean();
    }

    @Override
    public String handleMissingData(RowRecord record) {
        return mean.toString();
    }
}

File: src/main/scala/org/apache/datacommons/prepbuddy/cleansers/imputation/MeanSubstitution.scala
package org.apache.datacommons.prepbuddy.cleansers.imputation

import org.apache.datacommons.prepbuddy.rdds.TransformableRDD
import org.apache.datacommons.prepbuddy.utils.RowRecord

class MeanSubstitution extends ImputationStrategy {
    private var mean: Double = 0

    def handleMissingData(record: RowRecord): String = mean.toString

    def prepareSubstitute(rdd: TransformableRDD, missingDataColumn: Int): Unit = {
        mean = rdd.toDoubleRDD(missingDataColumn).mean()
    }
}
